### ISSUE TITLE
Unreadable jpeg metadata fields when including develop history

### ISSUE THREAD

**Issue by elstoc** [darktable-org/darktable#4355]
When exporting an image to jpeg, the 'Title' metadata field cannot be read by exiftool (command-line), libxmp (python) or flickr. When de-selecting 'develop history' in the 'edit metadata exportation' window (in the 'export selected' module), the issue disappears.

The Title field appears to be present in the file (since Geeqie displays it) but something is preventing other tools from reading it properly.

I've only been able to reproduce this issue on a single image so far and with exiftool/libxmp (possibly due to the fact that the image in question had a large history stack) but another user has also experienced this with flickr (see https://discuss.pixls.us/t/darktable-3-1-does-not-export-title-of-file/16589/3). 

Darktable version: 3.1.0+449~gc5f14d958-dirty
OS: ArchLinux
OpenCL: Issue occurs both with and without OpenCL

XMP attached

[2017-07-23-18.56.15.0_01.cr2.xmp.txt](https://github.com/darktable-org/darktable/files/4242474/2017-07-23-18.56.15.0_01.cr2.xmp.txt)

**Comment by johnny-bit** [comment]
Can you check and confirm whether #4357 fixed this for you?

**Comment by elstoc** [comment]
I've retested on latest master and the issue is still present

**Comment by phweyland** [comment]
> I've retested on latest master and the issue is still present

The xmp file seems to be ok. Could you give us access to the faulty jpeg file ?

**Comment by elstoc** [comment]
Here are two exports

The broken one (where develop history was included)
![2017-07-23-18 56 15 0_broken](https://user-images.githubusercontent.com/9555491/75551595-77037e80-5a2c-11ea-8823-8ef9a550512c.jpg)

The working one (where develop history was not included)
![2017-07-23-18 56 15 0_notbroken](https://user-images.githubusercontent.com/9555491/75551597-779c1500-5a2c-11ea-8f70-ec21ab4ae863.jpg)

**Comment by phweyland** [comment]
Strange. Not sure that will help.
the first has no xmp at all.
the second one has only the metadata with no development data.

```
[...]
Exif.Image.DateTimeOriginal                  Ascii      20  2017:07:23 18:56:15
Xmp.exif.DateTimeOriginal                    XmpText    19  2017:07:23 18:56:15
Xmp.xmp.Rating                               XmpText     1  2
Xmp.xmpMM.DerivedFrom                        XmpText    25  2017-07-23-18.56.15.0.cr2
Xmp.dc.title                                 LangAlt     1  lang="x-default" Ryde, Isle of Wight
Xmp.darktable.colorlabels                    XmpSeq      1  2
```

Could run darktable from console with option -d all and try to capture messages related to export ?
Maybe we would have some indications ...

**Comment by elstoc** [comment]
> the first has no xmp at all.

Presumably whatever the fault is, it's stopping both the history and some of the other metadata being extracted correctly.

> the second one has only the metadata with no development data

That's because this is the working one for comparison, where I explicitly chose not to include the develop history.

Here's the bit of the log that contains the export (darktable -d all). This is an export where I chose to include develop history (so the first example above)

[darktable4355.log](https://github.com/darktable-org/darktable/files/4268178/darktable4355.log)

**Comment by elstoc** [comment]
Here's the error:

`[xmp_attach] /home/chris/Photos/00 All Photos/000 Best Export/2017-07-23-18.56.15.0.jpg: caught exiv2 exception 'Size of XMP JPEG segment is larger than 65535 bytes'
`

**Comment by johnny-bit** [comment]
So... "store xmp tags in compressed format" (`compress_xmp_tags`) is disabled or simply not working in this case (as in works, but still can't store all the data)?

**Comment by elstoc** [comment]
"store xmp tags in compressed format" is enabled for this export

**Comment by johnny-bit** [comment]
so it seems that adding ~90kb of development data still exceeds max allowed for xmp segment...

I think darktable should detect that kind of situation and do something... what should darktable do in that situation (xmp data won't fit and we're certain of it)?

**Comment by elstoc** [comment]
Would it be possible to split it into multiple tags? The only other alternatives are to not include it at all, or try and extract a minimal subset of data (if that's even realistic). Probably needs another option in the xmp preferences section if we want the user to have a choice.  Could we prompt the user if/when it happens.

If we decide never to include it, probably needs at least a tooltip in the xmp parameter.

Not a big issue for me either way TBH. I'm happy never to include xmp data in my exports and will permanently switch this off now I know it's what caused the issue. Happy to test any fix though.

**Comment by johnny-bit** [comment]
There is no way (that I see) to split tags. Each tag in jpeg must fit in single APP container that's max 64kb, and there's max 16 of those uniquely-identified tags (if I read the specs correctly) and each has unique identifier...

My best current bet would be to calculate size of tags added and skip adding development history for those photos that would otherwise be left without xmp tags. since export is long-running task, popping up dialog or showing something to user is bad ux if it would mean breaking export flow.

So I think that best bet would be to warn user in options/export that for some files with large dev history, the history won't be saved to file. And then if such situation happens during export, simply put some info after completed export that some files haven't had history saved.

**Comment by elstoc** [comment]
I agree. That sounds like a reasonable approach.

**Comment by elstoc** [comment]
This error originates from a part of the code that ostensibly is trying to read the xmp/iptcData from the "Original File" in order to subsequently merge it with the darktable xmp data. However, I note that the error message above refers to an image "2017-07-23-18.56.15.0.jpg" which is the location of the **exported jpeg** and not the location of the input raw file. The code (`common/exif.cc` line 3587) appears to be getting this data by trying to call it up from the cache (`&from_cache` here is TRUE):

`    dt_image_full_path(imgid, input_filename, sizeof(input_filename), &from_cache);`

I don't really understand why it would be trying to extract metadata from a previously-exported copy of the file, from the cache. But then maybe I just don't understand what it's trying to do?

**Comment by Nilvus** [comment]
As I wrote future 3.2 article, I just see that this issue seems to have been fixed, at least by #4357 PR. @elstoc: could you confirm and if ok, close this one?

**Comment by elstoc** [comment]
I retested this shortly after #4357 was merged and it didn't fix the problem for me.

**Comment by johnny-bit** [comment]
This needs some love since it's exiv problem - a crash in exif during merging of history needs to be gracefully handled, so that tags are present but not necessarilly the develop history.

**Comment by phweyland** [comment]
> I retested this shortly after #4357 was merged and it didn't fix the problem for me.

I've made some tests. Nothing happens before dt tries to write the metadata chunk (all included).
What I will try then is, in that case, to relaunch the full export without history. Not very elegant however ...

**Comment by johnny-bit** [comment]
YEah, not really elegant...

But if there's no better way then working is better than pretty ;)

**Comment by phweyland** [comment]
I've another idea. If error is detected I can try to remove the dt history tags from the buffer.
@turbogit, Pascal if I remove "Xmp.darktable.history" am I done, or there are other tags which are associated to history to be removed too ?
 In fact I want to avoid further issues re-importing half an history ...

**Comment by TurboGit** [comment]
Philippe, I suppose you want also to remove:
- Xmp.darktable.masks_history
- Xmp.darktable.auto_presets_applied
- Xmp.darktable.history_end
- Xmp.darktable.iop_order_version
- Xmp.darktable.iop_order_list

**Comment by phweyland** [comment]
I need some help regarding c++. Sorry for the dumb question...

In dt_exif_xmp_attach_export()
img and xmpData are declared in the "try" block:
```
    std::unique_ptr<Exiv2::Image> img(Exiv2::ImageFactory::open(WIDEN(filename)));
    Exiv2::XmpData &xmpData = img->xmpData();
```
How to declare them in the upper block to access them from the "catch" block ?

**Comment by phweyland** [comment]
> How to declare them in the upper block to access them from the "catch" block ?

Got it. I'd started with the wrong one (xmpData). Starting with img is easier.
Thanks

**Comment by parafin** [comment]
You can't access them from the catch block, logically they might not have been created yet at the time of the exception. Most likely you need to create another try{}catch() around the function call you need to handle failure from. It would be easier if you told what you're trying to achieve.

**Comment by phweyland** [comment]
@parafin, thanks for responding.

What I'm trying to do is to remove history xmp data from xmpData buffer when I get the exception
"Size of XMP JPEG segment is larger than 65535 bytes"
with     `img->writeMetadata();`

I've put 
`  std::unique_ptr<Exiv2::Image> img = NULL;`
at the upper level in `dt_exif_xmp_attach_export()` and I can now access img (and then &xmpData) from the catch{} block.

Then I remove the history metadata from xmpData and retry 
`img->writeMetadata();`

But so far I get again the "Size of XMP JPEG segment is larger than 65535 bytes" exception ...

**Comment by parafin** [comment]
See my review on your PR, as I said you need to add another try-catch block.
